fullParseDice.py

```diff
@@ -0,0 +1,162 @@
+from random import randint
+
+'''
+Alright. First off, this file is clearly a hacked-together mess. The good news is that it works.
+The bad news that it takes a lot of print statements to figure out where the problem is if it breaks.
+
+in any case, for the inital call to simplifyExpression, pass in the roll statement to parse and a reference to an empty array. 
+'''
+def simplifyExpression(expression, dice, alreadyHashed = False, diceInstance = 0): #Parameters two and three are for # operations.
+    e = expression
+    eOld = ""
+    i=0 #solution. Set here in case of a non-# expression, which needs to change this value.
+    while True: #This loop can be broken out of. It will either end in an invalid string exception or a break.
+        if e.count(")") > 0 and (e.find("#") > e.find(")") or e.find("#") == -1): #Checking for parenthasis
+            rightSide = e.find(")") #Locates the leftmost right parenthasis
+            i = rightSide
+            leftSide = -1 #setting a dummy value
+            while i >= 0: #this loop just moves one character right repeatedly in search of a "("
+                if (e[i] == "("):
+                    leftSide = i+1
+                    break
+                i-=1
+            if leftSide == -1: #if no left parenthasis was found
+                raise Exception("Invalid String: Mismatched Parenthasis")
+            miniExpression = e[leftSide:rightSide]
+            miniSolution = simplifyExpression(miniExpression, dice, True, diceInstance) #Set to true to prevent hash inside of parenthasis
+            e = replaceInstance(e,str(miniSolution[0]),leftSide-1,rightSide+1) #[0] is because simplifyExpression returns an array of solutions, but there will never be a # in parenthasis. +/- 1 on sides is to also replace parenthasis
+
+        elif e.count("#") == 1 and not alreadyHashed: #There should never be more than one # in a statement
+            h = e.find("#") #Locates the #
+            leftbound = h
+            while leftbound > 0: #Same thing as above, just written later. It moves to the left until it hits the beginning of the string or find a non-number character.
+                if e[leftbound-1].isdigit():
+                    leftbound-=1
+                else:
+                    break
+                repetitions = int(e[leftbound:h]) #Once we know how many repetitions of the command we have to do...
+                i = []
+                for r in range(repetitions): #...we do the right side of the # as it's own expression, and return it.
+                    #KNOWN ISSUE: Ex: 3-2#1d20 will behave like 2#1d20. This can be mitigated by doing (3-2)#1d20 
+                    i.append(simplifyExpression(e[h+1:], dice, True, r+1)[0]) #puts all answers in an array.
+                return i #returns the array
```

Comment:
Please specify at the start of the function what is returned.

When making a function always detail what each parameter does and exactly what is returned.